### PR TITLE
Feature/aa 354 add api for remove accommodations with change country codes

### PR DESCRIPTION
**Pull request by hasmikharutyunyan**
Log will be added in the next PR, with new version of HappyTravel.LoggerMessageGenerator

### CODE REVIEW

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs [hunk 1]
```diff
@@ -27,6 +30,87 @@ public class AccommodationManagementService : IAccommodationManagementService
...
+                    {
+                        if (accommodation.SupplierAccommodationCodes.Count == 1)
+                        {
+                            _context.Remove(accommodation);
+                            removedHtIds.Add(accommodation.Id);
+                        }
+                        else
+                        {
+                            accommodation.SupplierAccommodationCodes.Remove(supplier);
+
```
Comment (mingazhev) [inline HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:96]: blank line can be removed

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs [hunk 2]
```diff
@@ -27,6 +30,87 @@ public class AccommodationManagementService : IAccommodationManagementService
...
 
 
+        public async Task<Result> RemoveDuplicatesFormedBySuppliersChangedCountry(List<Suppliers> suppliers)
+        {
+            foreach (var supplier in suppliers)
+            {
+                try
+                {
+                    var entityType = _context.Model.FindEntityType(typeof(RichAccommodationDetails))!;
+                    var tableName = entityType.GetTableName()!;
```
Comment (mingazhev) [inline HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:40]: May be it is better to move such low-level code to some helper class?

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs [hunk 3]
```diff
@@ -27,6 +30,87 @@ public class AccommodationManagementService : IAccommodationManagementService
...
+                            accommodation.SupplierAccommodationCodes.Remove(supplier);
+                          
+                            _context.Update(accommodation);
+                        }
+                    }
+
+                    await _context.SaveChangesAsync();
+                    await _accommodationChangePublisher.PublishRemoved(removedHtIds);
+                }
+                catch (Exception ex)
```
Comment (mingazhev) [inline HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:104]: Don't we need to log it?

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs (hunk 2 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:40]: Maybe, but this code is permanent, and we must remove that

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs (hunk 3 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:104]: I have written above that will add logs in the next pr

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs [hunk 4]
```diff
@@ -27,6 +30,87 @@ public class AccommodationManagementService : IAccommodationManagementService
...
+                    var countryCodesFromRawAccommodations = await _context.RawAccommodations
+                        .Where(ac => ac.Supplier == supplier
+                            && supplierCodes.Contains(ac.SupplierAccommodationId))
+                        .Select(ac => new
+                        {
+                            CountryCode = ac.CountryCode,
+                            SupplierCode = ac.SupplierAccommodationId
+                        }).ToListAsync();
+
+                    var parameters = new List<string>(supplierCodes) {supplierNameParameter};
```
Comment (kirillta) [inline HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:66]: This list isn't clear in terms of readability

File: HappyTravel.Nakijin.Api/Controllers/AccommodationManagementController.cs [hunk 5]
```diff
@@ -352,6 +352,24 @@ public async Task<IActionResult> MatchAccommodations(int sourceHtId, int htIdToM
...
+        {
+            var (_, isFailure, error) = await _accommodationManagementService.RemoveDuplicatesFormedBySuppliersChangedCountry(suppliers);
+            if (isFailure)
+                return BadRequest(error);
+
+            return Ok();
+        }
+
+
         private static CancellationTokenSource _accommodationDataMergeTokenSource =
```
[the remark this answers is not in the document]
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Api/Controllers/AccommodationManagementController.cs:373]: Only in AccommodationManagementController and LocationManagementController, and each token is specific for each controller

File: HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs [hunk 6]
```diff
@@ -27,6 +30,87 @@ public class AccommodationManagementService : IAccommodationManagementService
...
+                    var duplicateSupplierAccommodations = await _context.ProjectionsForGroupedAccommodations
+                        .FromSqlRaw(
+                            @$"SELECT a.""{columnName}""->> {{{0}}} as SupplierCode FROM ""{tableName}"" a 
+                                   WHERE a.""{columnName}""->> {{{0}}} IS NOT NULL AND a.""IsActive""
+                                   GROUP BY a.""{columnName}""->> {{{0}}}
+                                   HAVING count(a.""{columnName}""->> {{{0}}}) > 1",
+                            new object[] {supplierNameParameter})
+                        .ToListAsync();
+
+                    var supplierCodes = duplicateSupplierAccommodations.Select(ac => ac.SupplierCode).ToList();
```
[the remark this answers is not in the document]
Comment (hasmikharutyunyan) [reply HappyTravel.Nakijin.Api/Services/AccommodationManagementService.cs:55]: I have tried, and I get the error

**Review by mingazhev — APPROVED** [review]

**Review by kirillta — APPROVED** [review]

**Pull request merged**